server/service/src/document/Programs.md

```diff
@@ -0,0 +1,144 @@
+
+## Derive custom JSON schema from base schema
+
+Common documents types have predefined commonly understood JSON schemas.
+To extend a base schema new fields can be added to the base schema in a backwards compatible way.
+Only adding new fields to a schema ensures that derived schemas still fullfil the base schema and thus can be handle generically.
+For example:
+
+```json
+{
+  // (base schema would need to be inlined though...)
+  "allOf": [{ "$ref": "/schema/base" }],
+  "properties": {
+    "baz": { "type": "string" }
+  }
+}
+```
+
+## Document registry
+
+In general documents are related to other documents, e.g. an encounter document is related to a certain type of program enrolment document.
+These relations of document types are stored in the document registry.
+The document registry also stores the json schema for a document type.
+For example, when editing a new document of a certain type the document registry dictates which json schema the document data must follow.
+
+Furthermore, the document registry contains a ui schema which defines how a document should be displayed.
+Currently there is exactly one ui schema for each document type but in the future it might be necessary to support multiple ui schema depending on the context in which a document is shown.
+For example, there could be a UI schema for the full document and one that only gives a summary of the document.
+
+## Document versioning
+
+Each edit of a document is stored as a new version.
+The `document.id` is the version id while `document.name` is the general document identifier to shared between all document versions.
+TODO: rename `document.id` => `document.versionId` and `document.name` => `document.id`?
+
+# Customizable Program User Interface
+
+One of the requirements of the programs module is to provide a customizable user interface to display and edit custom document types.
+To achieve this JSONForms UI schema files are used to define a customized UI.
+JSONForms give UI designer great flexibility on how document data is composed and displayed.
+Furthermore, custom control can be created to implemented specialized UI elements such as graphs or controls to generate patient IDs.
+
+## Event trigger Controls
+
+There some specialized controls which add events to the document data based on some user defined triggers.
+
+## None standard controls and options
+
+Specialized controls are located in `packages/system/src/Patient/JsonForms/components`.
+These component files also contain type information and js docs for the component options.
+
+## Encounters data extraction
+
+The graph control needs to fetch data from multiple encounters.
+This is done through a fairly generic `encounterFields` endpoint which allows to extract an field from an encounter document.
+For example, it allows to fetch all `data.physicalExamination.weight` fields from all patient encounters (note, that the weight is a custom field).
```

Comment:
```suggestion
This is done through a fairly generic `encounterFields` endpoint which allows the field to be extracted from an encounter document.
For example, the `data.physicalExamination.weight` field from all patient encounters can be fetched (note, that the weight is a custom field).
```

Reply:
Now:

This is done through a fairly generic `encounterFields` endpoint which allows users to extract fields from an encounter document.
For example, the `data.physicalExamination.weight` field can be fetched from all patient encounters (note, that the weight is a custom field).